File: HandLazy app/handlazy/android/app/src/main/java/com/example/handlazy/GestureAccessibilityService.java
package com.example.handlazy;

import android.accessibilityservice.AccessibilityService;
import android.accessibilityservice.GestureDescription;
import android.graphics.Path;
import android.os.Handler;
import android.os.Looper;
import android.view.accessibility.AccessibilityEvent;
import android.util.Log;

public class GestureAccessibilityService extends AccessibilityService {
    private static final String TAG = "HandLazyGesture";
    private static GestureAccessibilityService instance;
    private android.view.WindowManager windowManager;
    private android.view.View cursorView;
    private android.view.WindowManager.LayoutParams cursorParams;
    private Handler mainHandler;

    @Override
    public void onCreate() {
        super.onCreate();
        instance = this;
        mainHandler = new Handler(Looper.getMainLooper());
        windowManager = (android.view.WindowManager) getSystemService(WINDOW_SERVICE);
        createCursorView();
        Log.d(TAG, "Accessibility Service Created");
    }

    @Override
    public void onAccessibilityEvent(AccessibilityEvent event) {
        // Not needed for gesture performance
    }

    @Override
    public void onInterrupt() {
        Log.d(TAG, "Service Interrupted");
    }

    private void createCursorView() {
        cursorView = new android.view.View(this);
        android.graphics.drawable.GradientDrawable drawable = new android.graphics.drawable.GradientDrawable();
        drawable.setShape(android.graphics.drawable.GradientDrawable.OVAL);
        drawable.setColor(android.graphics.Color.YELLOW);
        drawable.setStroke(3, android.graphics.Color.RED);
        cursorView.setBackground(drawable);

        cursorParams = new android.view.WindowManager.LayoutParams(
            40, // Width
            40, // Height
            android.view.WindowManager.LayoutParams.TYPE_ACCESSIBILITY_OVERLAY,
            android.view.WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE |
            android.view.WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE |
            android.view.WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN |
            android.view.WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS,
            android.graphics.PixelFormat.TRANSLUCENT
        );
        cursorParams.gravity = android.view.Gravity.TOP | android.view.Gravity.START;
        cursorParams.x = 0;
        cursorParams.y = 0;
    }

    public void showCursor() {
        if (cursorView.getParent() == null) {
            mainHandler.post(() -> {
                try {
                    windowManager.addView(cursorView, cursorParams);
                } catch (Exception e) {
                    Log.e(TAG, "Error showing cursor: " + e.getMessage());
                }
            });
        }
    }

    public void hideCursor() {
        if (cursorView.getParent() != null) {
            mainHandler.post(() -> {
                try {
                    windowManager.removeView(cursorView);
                } catch (Exception e) {
                    Log.e(TAG, "Error hiding cursor: " + e.getMessage());
                }
            });
        }
    }

    public void updateCursorPosition(float x, float y) {
        if (cursorView.getParent() == null) {
             showCursor();
        }
        
        mainHandler.post(() -> {
            try {
                int screenWidth = getResources().getDisplayMetrics().widthPixels;
                int screenHeight = getResources().getDisplayMetrics().heightPixels;

                // X is mirrored (1.0 - x) because front camera matches mirror logic usually,
                // but depends on how flutter sends it. Let's assume Flutter sends raw landmark.
                // Raw MediaPipe landmarks: x is 0 on left, 1 on right of IMAGE.
                // Front camera image is usually mirrored relative to user?
                // Let's stick to standard mapping first.
                
                cursorParams.x = (int) (x * screenWidth);
                cursorParams.y = (int) (y * screenHeight);
                windowManager.updateViewLayout(cursorView, cursorParams);
            } catch (Exception e) {
                // Ignore transient update errors
            }
        });
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        hideCursor();
        instance = null;
        Log.d(TAG, "Accessibility Service Destroyed");
    }

    public static GestureAccessibilityService getInstance() {
        return instance;
    }

    public static boolean isServiceEnabled() {
        return instance != null;
    }

    // Perform swipe up gesture (Next Reel)
    public void swipeUp() {
        showOneTimeFeedback(android.graphics.Color.GREEN); // Feedback flash
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
             int screenHeight = getResources().getDisplayMetrics().heightPixels;
             int screenWidth = getResources().getDisplayMetrics().widthPixels;

             Path swipePath = new Path();
             swipePath.moveTo(screenWidth / 2f, screenHeight * 0.7f);
             swipePath.lineTo(screenWidth / 2f, screenHeight * 0.3f);

             GestureDescription.Builder gestureBuilder = new GestureDescription.Builder();
             gestureBuilder.addStroke(new GestureDescription.StrokeDescription(swipePath, 0, 300));

             dispatchGesture(gestureBuilder.build(), null, null);
        }
    }

    // Perform swipe down gesture (Previous Reel)
    public void swipeDown() {
        showOneTimeFeedback(android.graphics.Color.MAGENTA); // Feedback flash
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
            int screenHeight = getResources().getDisplayMetrics().heightPixels;
            int screenWidth = getResources().getDisplayMetrics().widthPixels;

            Path swipePath = new Path();
            swipePath.moveTo(screenWidth / 2f, screenHeight * 0.3f);
            swipePath.lineTo(screenWidth / 2f, screenHeight * 0.7f);

            GestureDescription.Builder gestureBuilder = new GestureDescription.Builder();
            gestureBuilder.addStroke(new GestureDescription.StrokeDescription(swipePath, 0, 300));

            dispatchGesture(gestureBuilder.build(), null, null);
        }
    }
    
    // Quick visual flash for action feedback
    private void showOneTimeFeedback(int color) {
        if (cursorView.getParent() != null) {
            mainHandler.post(() -> {
                 android.graphics.drawable.GradientDrawable bg = (android.graphics.drawable.GradientDrawable) cursorView.getBackground();
                 bg.setColor(color);
                 cursorView.invalidate();
                 
                 // Revert to yellow after 500ms
                 mainHandler.postDelayed(() -> {
                     bg.setColor(android.graphics.Color.YELLOW);
                     cursorView.invalidate();
                 }, 500);
            });
        }
    }

    // Set volume (0-100)
    public void setVolume(int volumePercent) {
        showOneTimeFeedback(android.graphics.Color.CYAN); // Feedback flash
        android.media.AudioManager audioManager = 
            (android.media.AudioManager) getSystemService(AUDIO_SERVICE);
        if (audioManager != null) {
            int maxVolume = audioManager.getStreamMaxVolume(android.media.AudioManager.STREAM_MUSIC);
            int volume = (int) (maxVolume * volumePercent / 100.0);
            audioManager.setStreamVolume(
                android.media.AudioManager.STREAM_MUSIC, 
                volume, 
                0
            );
            Log.d(TAG, "Volume set to: " + volumePercent + "%");
        }
    }
}

File: HandLazy app/handlazy/android/app/src/main/kotlin/com/example/handlazy/GestureAccessibilityService.kt
package com.example.handlazy

import android.accessibilityservice.AccessibilityService
import android.accessibilityservice.GestureDescription
import android.content.res.Resources
import android.graphics.Color
import android.graphics.Path
import android.graphics.PixelFormat
import android.graphics.drawable.GradientDrawable
import android.media.AudioManager
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.Gravity
import android.view.View
import android.view.WindowManager
import android.view.accessibility.AccessibilityEvent
import com.example.handlazy.math.CoordinateMapper

/**
 * Accessibility Service for HandLazy that:
 * 1. Displays a persistent yellow dot cursor overlay
 * 2. Dispatches system gestures (scroll, click) for reel navigation
 */
class GestureAccessibilityService : AccessibilityService() {

    companion object {
        private const val TAG = "HandLazy"
        
        @Volatile
        private var instance: GestureAccessibilityService? = null
        
        fun getInstance(): GestureAccessibilityService? = instance
        
        fun isServiceEnabled(): Boolean = instance != null
    }

    private lateinit var windowManager: WindowManager
    private lateinit var coordinateMapper: CoordinateMapper
    private var cursorView: View? = null
    private var cursorParams: WindowManager.LayoutParams? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    
    // Cursor size
    private val cursorSize = 40 // dp
    
    override fun onServiceConnected() {
        super.onServiceConnected()
        instance = this
        
        windowManager = getSystemService(WINDOW_SERVICE) as WindowManager
        
        // Initialize coordinate mapper with screen dimensions
        val metrics = Resources.getSystem().displayMetrics
        coordinateMapper = CoordinateMapper(metrics.widthPixels, metrics.heightPixels)
        
        Log.d(TAG, "Accessibility Service Connected. Screen: ${metrics.widthPixels}x${metrics.heightPixels}")
    }

    override fun onAccessibilityEvent(event: AccessibilityEvent?) {
        // Not needed for gesture control
    }

    override fun onInterrupt() {
        Log.d(TAG, "Service Interrupted")
    }

    override fun onDestroy() {
        super.onDestroy()
        hideCursor()
        instance = null
        Log.d(TAG, "Accessibility Service Destroyed")
    }

    // ==================== CURSOR OVERLAY (DISABLED for v13.0) ====================
    // User requested floating camera preview instead of yellow dot.
    
    private fun createCursorView(): View {
        return View(this) // Dummy
    }

    fun showCursor() {
        // Disabled
    }

    fun hideCursor() {
        // Disabled
    }

    fun updateCursorPosition(normX: Float, normY: Float) {
        // Disabled
    }
    
    fun flashCursor(color: Int) {
        // Disabled
    }

    // ==================== GESTURE ACTIONS ====================
    
    /**
     * Perform swipe up gesture (Next Reel).
     */
    fun swipeUp() {
        flashCursor(Color.GREEN)
        
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            val metrics = Resources.getSystem().displayMetrics
            val screenHeight = metrics.heightPixels
            val screenWidth = metrics.widthPixels

            val path = Path().apply {
                moveTo(screenWidth / 2f, screenHeight * 0.7f)
                lineTo(screenWidth / 2f, screenHeight * 0.3f)
            }

            val gesture = GestureDescription.Builder()
                .addStroke(GestureDescription.StrokeDescription(path, 0, 300))
                .build()

            dispatchGesture(gesture, object : GestureResultCallback() {
                override fun onCompleted(gestureDescription: GestureDescription?) {
                    Log.d(TAG, "Swipe UP completed - Next Reel")
                }
                override fun onCancelled(gestureDescription: GestureDescription?) {
                    Log.w(TAG, "Swipe UP cancelled")
                }
            }, null)
        }
    }

    /**
     * Perform swipe down gesture (Previous Reel).
     */
    fun swipeDown() {
        flashCursor(Color.MAGENTA)
        
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            val metrics = Resources.getSystem().displayMetrics
            val screenHeight = metrics.heightPixels
            val screenWidth = metrics.widthPixels

            val path = Path().apply {
                moveTo(screenWidth / 2f, screenHeight * 0.3f)
                lineTo(screenWidth / 2f, screenHeight * 0.7f)
            }

            val gesture = GestureDescription.Builder()
                .addStroke(GestureDescription.StrokeDescription(path, 0, 300))
                .build()

            dispatchGesture(gesture, object : GestureResultCallback() {
                override fun onCompleted(gestureDescription: GestureDescription?) {
                    Log.d(TAG, "Swipe DOWN completed - Prev Reel")
                }
                override fun onCancelled(gestureDescription: GestureDescription?) {
                    Log.w(TAG, "Swipe DOWN cancelled")
                }
            }, null)
        }
    }

    /**
     * Set system media volume (0-100).
     */
    fun setVolume(volumePercent: Int) {
        flashCursor(Color.CYAN)
        
        val audioManager = getSystemService(AUDIO_SERVICE) as? AudioManager
        audioManager?.let { am ->
            val maxVolume = am.getStreamMaxVolume(AudioManager.STREAM_MUSIC)
            val volume = (maxVolume * volumePercent / 100.0).toInt()
            am.setStreamVolume(AudioManager.STREAM_MUSIC, volume, 0)
            Log.d(TAG, "Volume set to: $volumePercent%")
        }
    }
}
